go/cmd/dolt/commands/migrate.go

```diff
@@ -45,31 +55,218 @@ func (cmd MigrateCmd) CreateMarkdown(_ filesys.Filesys, _, _ string) error {
+	ap := cmd.createArgParser()
+	help, _ := cli.HelpAndUsagePrinters(cli.GetCommandDocumentation(commandStr, pushDocs, ap))
+	apr := cli.ParseArgs(ap, args, help)
+
+	if apr.Contains(migratePushFlag) && apr.Contains(migratePullFlag) {
+		cli.PrintErrf(color.RedString("options --%s and --%s are mutually exclusive", migratePushFlag, migratePullFlag))
 		return 1
 	}
-	if !needed {
-		cli.Println("Repository format is up to date")
-		return 0
+
+	var err error
+	switch {
+	case apr.Contains(migratePushFlag):
+		err = pushMigratedRepo(ctx, dEnv, apr)
+	case apr.Contains(migratePullFlag):
+		err = fetchMigratedRemoteBranches(ctx, dEnv, apr)
+	default:
+		err = migrateLocalRepo(ctx, dEnv)
 	}
 
-	cli.Println(color.YellowString(migrationMsg))
-	err = rebase.MigrateUniqueTags(ctx, dEnv)
 	if err != nil {
-		cli.PrintErrf("error migrating repository: %s", err.Error())
+		cli.PrintErrln(color.RedString(err.Error()))
 		return 1
 	}
+
 	return 0
 }
 
+func migrateLocalRepo(ctx context.Context, dEnv *env.DoltEnv) error {
+	localMigrationNeeded, err := rebase.NeedsUniqueTagMigration(ctx, dEnv.DoltDB)
+
+	if err != nil {
+		return err
+	}
+
+	if localMigrationNeeded {
+		cli.Println(color.YellowString(migrationMsg))
+		err = rebase.MigrateUniqueTags(ctx, dEnv)
+
+		if err != nil {
+			return err
+		}
+	} else {
+		cli.Println("Repository format is up to date")
+	}
+
+	remoteName := "origin"
+	remoteMigrated, err := remoteHasBeenMigrated(ctx, dEnv, remoteName)
+	if err != nil {
+		// if we can't check the remote, exit silently
+		return nil
+	}
+
+	if !remoteMigrated {
+		cli.Println(fmt.Sprintf("Remote %s has not been migrated", remoteName))
+		cli.Println(fmt.Sprintf("Run 'dolt migrate --push' %s to update remote", remoteName))
+	} else {
+		cli.Println(fmt.Sprintf("Remote %s has been migrated", remoteName))
+		cli.Println(fmt.Sprintf("Run 'dolt migrate --pull %s' to update refs", remoteName))
+	}
+
+	return nil
+}
+
+func pushMigratedRepo(ctx context.Context, dEnv *env.DoltEnv, apr *argparser.ArgParseResults) error {
+	localMigrationNeeded, err := rebase.NeedsUniqueTagMigration(ctx, dEnv.DoltDB)
+	if err != nil {
+		return err
+	}
+	if localMigrationNeeded {
+		return fmt.Errorf("Local repo must be migrated before pushing, run 'dolt migrate'")
+	}
+
+	remoteName := "origin"
+	if apr.NArg() > 0 {
+		remoteName = apr.Arg(0)
+	}
+
+	remotes, err := dEnv.GetRemotes()
+	if err != nil {
+		return err
+	}
+
+	remote, remoteOK := remotes[remoteName]
+	if !remoteOK {
+		return fmt.Errorf("unknown remote %s", remoteName)
+	}
+
+	remoteMigrated, err := remoteHasBeenMigrated(ctx, dEnv, remoteName)
+	if err != nil {
+		return err
+	}
+	if remoteMigrated {
+		return fmt.Errorf("Remote %s has been migrated\nRun 'dolt migrate --pull' to update refs", remoteName)
+	} else {
+		// force push all branches
+		bb, err := dEnv.DoltDB.GetBranches(ctx)
+
+		if err != nil {
+			return err
+		}
+
+		for _, branch := range bb {
```

Comment:
In general this seems weird to me from a UX perspective. Maybe not super important at this point...ideally I guess we would be pushing the remote refs that actually exist, instead  of just creating a remote ref for every local branch we happen to have.